### PR TITLE
Add Client Status Discovery Service (CSDS) API definition

### PR DESCRIPTION
**Pull request by fuqianggao**
Add Client Status Discovery Service (CSDS) API definition. This can be used by debug tools to obtain config information for specific clients from control plane. 

Risk Level: Low
Testing: N/A
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by jmarantz** [comment]
@fuqianggao can you look at CI?

**Comment by fuqianggao** [comment]
> @fuqianggao can you look at CI?

Done.

**Comment by mattklein123** [comment]
@fuqianggao thanks for this. This is interesting. To help with review, can you provide some more high level context on rationale, etc.? Thank you.

/wait-any

**Comment by fuqianggao** [comment]
@mattklein123 The purpose of this API is to provide a way for control plane to expose its view on the status of clients connected to it. For example, in a service mesh with many Envoys, a standalone debug tool can connect to control plane and stream CSDS, to know which Envoys are actually connected to the control plane and what configs they are getting. This will be very helpful for debugging purpose. Istio has a similar command line tool: https://istio.io/docs/ops/diagnostic-tools/proxy-cmd/, I imagine they can possibly use CSDS. Besides configs, it's also possible to include other stats in CSDS, but there are other means to expose those (e.g. stackdriver).

**Comment by mattklein123** [comment]
@fuqianggao agreed this is a very useful thing to add, thanks for the context. I'm tagging a bunch of people at Lyft who will be interested in this as we are looking at building out some behavior along these lines in a management tool.

cc @danielhochman @ryancox @sindhura

File: api/envoy/service/discovery/v2/csds.proto [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    admin.v2alpha.RoutesConfigDump route_config = 4;
+
+    admin.v2alpha.ScopedRoutesConfigDump scoped_route_config = 5;
+  }
+}
+
+// All xds configs for a particular client.
+message ClientConfig {
+  api.v2.core.Node node = 1;
```
Comment (htuch) [inline api/envoy/service/discovery/v2/csds.proto]: Needs documentation here.

File: api/envoy/service/discovery/v2/csds.proto [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  // Config is not sent.
+  NOT_SENT = 2;
+
+  // Management server has sent the config to client but hasn’t received
+  // ACK/NACK.
+  STALE = 3;
+
+  // Management server has sent the config to client but received NACK.
+  ERROR = 4;
+}
```
Comment (htuch) [inline api/envoy/service/discovery/v2/csds.proto]: One question I still have is that status such as stale is useful to obtain, and you want to know the config responsible for this, but do you also want to know the previous synced config? Would CSDS report both?

Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: That's a good question. We can potentially add a last_synced_config field. But management server might not have the last_synced_config available if it doesn't keep track of older configs.

File: api/envoy/service/discovery/v2/csds.proto (hunk 1 above)
Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: Done.

File: api/envoy/service/discovery/v2/csds.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/service/discovery/v2/csds.proto]: Could you share both current and stale configs in the config list? With timestamps it should be possible to figure out which is which.

File: api/envoy/service/discovery/v2/csds.proto [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+option java_generic_services = true;
+
+import "envoy/admin/v2alpha/config_dump.proto";
+import "envoy/api/v2/core/base.proto";
+
+import "google/api/annotations.proto";
+import "google/protobuf/struct.proto";
+
+// CSDS is Client Status Discovery Service. It can be used to get the status of
+// xDS-compliant client from the management server's point of view.
```
Comment (mattklein123) [inline api/envoy/service/discovery/v2/csds.proto]: When I had originally skimmed this PR, one of the things that I thought this might cover is a more systematic way of fetching the current state *from the client*. So basically a formalized version of config dump. This would allow a "client tracking server" to potentially get updates from Envoy, etc. on its current status to help with reconciliation. Is this something you have thought about? It's not the same as what is being proposed here but is closely related IMO.

File: api/envoy/service/discovery/v2/csds.proto (hunk 2 above)
Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: Good idea. In that case, no api change needed. It's all up to the management server what configs to send back.

File: api/envoy/service/discovery/v2/csds.proto (hunk 3 above)
Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: Thanks! Yes, I have thought about that and had some discussion with htuch@. If it's just configs, we could re-use this API to fetch configs from Envoy as well. However, if we also want to expose stats, then the stats from control plane and from Envoy can be quite different, and it would be confusing to use the same API.
Alternatively, we could have another API that's basically Admin Discovery Service, where Envoy can expose it's admin interface (configs, stats).  WDYT?

Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: @htuch for further inputs.

Comment (mattklein123) [reply api/envoy/service/discovery/v2/csds.proto]: I think there are a bunch of ways we could go here per your comment. I'm CCing @sindhura who is working on a system at Lyft which right now going to fetch most data directly from the admin server. But I think it would be great if there were a more formalized way of accomplishing this. Hopefully she can provide some thoughts (probably delayed due to holidays).

Comment (htuch) [reply api/envoy/service/discovery/v2/csds.proto]: I think it really depends on what we want to be fetching. I think some parts of the admin interface are sort of covered by the idea of a `ConfigDump` style API. This could be extended to capture something closer to the dynamic aspect of `/clusters` (e.g. with health status).

Sometime stats comes up in these conversations, I think stats is really something that shouldn't be handled at all by some new API; we have an existing extension point and ways of moving stats out, the admin endpoint here is just a convenience.

Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: I have updated the comments to mention that this can also be used as an interface to fetch client state. PTAL. Thanks!

**Comment by howardjohn** [comment]
related work on istio side: https://github.com/istio/api/issues/1213

tl;dr is doing a similar thing but outside of envoy

**Comment by htuch** [comment]
@fuqianggao if it's all the same to you (i.e. no rush), I think we should hold back on this one until v3 lands. We should be doing the major v3 upgrade commits in the coming days.

**Comment by mattklein123** [comment]
Can you merge master and fix format and we can take another pass?

/wait

File: api/envoy/service/discovery/v2/csds.proto [hunk 4]
```diff
@@ -0,0 +1,86 @@
+syntax = "proto3";
+
+package envoy.service.discovery.v2;
+
+option java_package = "io.envoyproxy.envoy.service.discovery.v2";
+option java_outer_classname = "CsdsProto";
+option java_multiple_files = true;
+option java_generic_services = true;
+
```
Comment (htuch) [inline api/envoy/service/discovery/v2/csds.proto]: This should live in `envoy.service.status.v2` as per the new package layout https://github.com/envoyproxy/envoy/pull/9555

File: api/envoy/service/discovery/v2/csds.proto [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import "envoy/api/v2/core/base.proto";
+
+import "google/api/annotations.proto";
+import "google/protobuf/struct.proto";
+
+// CSDS is Client Status Discovery Service. It can be used to get the status of
+// xDS-compliant client from the management server's point of view. It can also
+// be used as an interface to get the current state of the client.
+service ClientStatusService {
+  rpc StreamClientStatus(stream ClientStatusRequest) returns (stream ClientStatusResponse) {
```
Comment (htuch) [inline api/envoy/service/discovery/v2/csds.proto]: Will need resource type annotations as soon as #9566 lands.

Comment (htuch) [inline api/envoy/service/discovery/v2/csds.proto]: Why is this not DiscoveryRequest/Response? I though this was a first-class xDS protocol.

**Comment by junr03** [comment]
@fuqianggao there are other commits in your branch. Do you mind cleaning up the commit history so that reviewers have an easier time? Thanks!

**Comment by fuqianggao** [comment]
Sorry, please don't review. I will clean up.

File: api/envoy/service/discovery/v2/csds.proto (hunk 4 above)
Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: Done.

File: api/envoy/service/discovery/v2/csds.proto (hunk 5 above)
Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: As discussed offline, the client status request can include a list of clients. Would like to keep the current definition.

Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: Done.

**Comment by fuqianggao** [comment]
@mattklein123 @htuch, I have moved csds under envoy/service/status/ to follow the new structure. PTAL. Thanks!

File: api/envoy/service/status/v2/csds.proto [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  STALE = 3;
+
+  // Management server has sent the config to client but received NACK.
+  ERROR = 4;
+}
+
+// Request for client status for a list of clients identified by Node identifier.
+message ClientStatusRequest {
+  // Node identifier for a list of clients.
+  repeated api.v2.core.Node nodes = 1;
```
Comment (htuch) [inline api/envoy/service/status/v2/csds.proto]: The more I think about this, the more I don't like using `Node` here. `Node` has a ton of information in principle, including build version, extension info, client features. It is designed to be used to communicate actual information about an Envoy node as well as identify it. We should just use the node ID IMHO.

File: api/envoy/service/status/v2/csds.proto [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+  // Metadata is used to give extra information to the management server.
+  // For example, management server can choose to only send config status if
+  // there is a metadata CONFIG_STATUS_ONLY: true.
+  google.protobuf.Struct metadata = 2;
```
Comment (htuch) [inline api/envoy/service/status/v2/csds.proto]: Not a fan of just adding an unstructured bag here, as we then need to standardize on the key namespace. Can we just make this a set of options that live as first class entries in the API?

File: api/envoy/service/status/v2/csds.proto (hunk 6 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: Done.

File: api/envoy/service/status/v2/csds.proto (hunk 7 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: The reason I put it as metadata is that I think it is hard to standardize some of the options. For my use case, I will need project number and network name. However, I am not sure those are needed when used by other control plane. WDYT?

Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: The reason I put it as metadata is that I think it is hard to standardize some of the options. For my use case, I will need project number and network name. However, I am not sure those are needed when used by other control plane. WDYT?

File: api/envoy/service/status/v2/csds.proto (hunk 6 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: Done.

**Comment by rshriram** [comment]
It seems a little cumbersome to have to specify all the node ids especially if this is user facing api.
fwiw, we are evolving our simple proxy command you cited (see https://github.com/istio/api/pull/1213/)

**Comment by fuqianggao** [comment]
> It seems a little cumbersome to have to specify all the node ids especially if this is user facing api.
> fwiw, we are evolving our simple proxy command you cited (see [istio/api#1213](https://github.com/istio/api/pull/1213))

Ack. We do need the node id to identify which config belongs to which client.

File: api/envoy/service/status/v2/csds.proto (hunk 7 above)
Comment (htuch) [reply api/envoy/service/status/v2/csds.proto]: Well, the question is how do client and server agree on this in a standard way? How do you plan on coordinating this with multiple client and server implementations?

File: api/envoy/service/status/v2/csds.proto [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    admin.v2alpha.RoutesConfigDump route_config = 4;
+
+    admin.v2alpha.ScopedRoutesConfigDump scoped_route_config = 5;
+  }
+}
+
+// All xds configs for a particular client.
+message ClientConfig {
+  // Node identifier for a particular client.
```
Comment (htuch) [inline api/envoy/service/status/v2/csds.proto]: I think you do want to return the actual `Node` here. The management server knows the full `Node` as reported by the client, so that seems a useful debug info.

File: api/envoy/service/status/v2/csds.proto (hunk 7 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: It's hard. Because information needed by different server impl could be different. For example, project number and network name makes sense for a control plane running in cloud, but could have no real meaning for a control plane running on-prem. Maybe, for now, we can expose project number and network name as options for cloud use cases, and later, add options that's relevant to on-prem and such? WDYT?

Comment (htuch) [reply api/envoy/service/status/v2/csds.proto]: How about expressing this as a `Node` metadata matcher? Would that be sufficient? Then you can express things like project ID as Node metadata and match on them when the client presents. We already have various places we match in the API, e.g. subset LB.

File: api/envoy/service/status/v2/csds.proto (hunk 8 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: Done.

File: api/envoy/service/discovery/v2/csds.proto (hunk 3 above)
Comment (fuqianggao) [reply api/envoy/service/discovery/v2/csds.proto]: Updated. PTAL.

File: api/envoy/service/status/v2/csds.proto [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+  // Management server has sent the config to client but received NACK.
+  ERROR = 4;
+}
+
+// Request for client status of clients identified by a list of NodeMatchers.
+message ClientStatusRequest {
+  message NodeMatcher {
+    // Specifies match criteria on the client node id.
+    type.matcher.StringMatcher node_id_matcher = 1;
```
Comment (htuch) [inline api/envoy/service/status/v2/csds.proto]: Nit: I think that because these matchers are already inside a matcher object, with field_name `node_matchers`, we can probably drop the `_matcher` suffix from the fields.

File: api/envoy/service/status/v2/csds.proto [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+option java_outer_classname = "CsdsProto";
+option java_multiple_files = true;
+option java_generic_services = true;
+
+// CSDS is Client Status Discovery Service. It can be used to get the status of
+// xDS-compliant client from the management server's point of view. In the
+// future, it can also potentially be used as an interface to get the current
+// state directly from the client.
+service ClientStatusService {
+  option (envoy.annotations.resource).type = "envoy.service.status.v2.ClientStatus";
```
Comment (htuch) [inline api/envoy/service/status/v2/csds.proto]: This isn't needed if you're not using `Discovery{Request,Response}`.

File: api/envoy/service/status/v2/csds.proto (hunk 7 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: Done. Thanks!

File: api/envoy/service/status/v2/csds.proto (hunk 9 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: Done.

File: api/envoy/service/status/v2/csds.proto (hunk 10 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: Done.

**Comment by htuch** [comment]
LGTM but you need to link this into the doc tree (see the docs CI failure).

@mattklein123 do you want to do a non-Google API pass, since both @fuqianggao and I are Googlers?

**Comment by mattklein123** [comment]
Can you merge master and I can take a final pass. Thank you.

/wait

**Comment by mattklein123** [comment]
Can you check format CI?

/wait

**Comment by fuqianggao** [comment]
> Can you check format CI?
> 
> /wait

I fails with: fatal: reference is not a tree, not sure how to fix it. Any suggestions? @mattklein123 @htuch

**Comment by htuch** [comment]
I've been hitting this as well with https://github.com/envoyproxy/envoy/pull/9706/checks?check_run_id=394140837. Maybe some GitHub mirroring or consistency issue?

**Comment by fuqianggao** [comment]
@mattklein123 @htuch The issues seems to be gone. CI is passing now. PTAL. Thanks!

File: api/envoy/service/status/v2/csds.proto [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+// Request for client status of clients identified by a list of NodeMatchers.
+message ClientStatusRequest {
+  message NodeMatcher {
+    // Specifies match criteria on the client node id.
+    type.matcher.StringMatcher node_id = 1;
+
+    // Specifies match criteria on the client node metadata.
+    type.matcher.MetadataMatcher node_metadata = 2;
+  }
```
Comment (mattklein123) [inline api/envoy/service/status/v2/csds.proto]: Can you go ahead and move this out into the types path as an independent matcher, along with the other matchers? I know of another specific use case in which we are going to need to do node matching, and I'm sure there are others that will come up soon. cc @jessicayuen for XDS aggregator overlap on the node matching config.

File: api/envoy/service/status/v2/csds.proto [hunk 12]
```diff
@@ -0,0 +1,92 @@
+syntax = "proto3";
+
+package envoy.service.status.v2;
```
Comment (mattklein123) [inline api/envoy/service/status/v2/csds.proto]: Should we start this as a v2alpha/v3alpha API or do you want to go straight to stable?

File: api/envoy/service/status/v2/csds.proto [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+  ConfigStatus status = 1;
+
+  oneof per_xds_config {
+    admin.v2alpha.ListenersConfigDump listener_config = 2;
+
+    admin.v2alpha.ClustersConfigDump cluster_config = 3;
+
+    admin.v2alpha.RoutesConfigDump route_config = 4;
+
+    admin.v2alpha.ScopedRoutesConfigDump scoped_route_config = 5;
```
Comment (mattklein123) [inline api/envoy/service/status/v2/csds.proto]: Do these need to be Anys in the case of mixed v2/v3, similar to the problem that @htuch just faced on config dump? cc @htuch

Comment (htuch) [reply api/envoy/service/status/v2/csds.proto]: Nup, the `*ConfigDump` protos already have the `Any` messages inside them.

File: api/envoy/service/status/v2/csds.proto (hunk 12 above)
Comment (htuch) [reply api/envoy/service/status/v2/csds.proto]: My take is we go straight to stable, this is a pretty straightforward API surface.

Comment (mattklein123) [reply api/envoy/service/status/v2/csds.proto]: OK sounds good.

File: api/envoy/service/status/v2/csds.proto (hunk 13 above)
Comment (mattklein123) [reply api/envoy/service/status/v2/csds.proto]: :+1:

File: api/envoy/service/status/v2/csds.proto (hunk 11 above)
Comment (fuqianggao) [reply api/envoy/service/status/v2/csds.proto]: Done.

**Comment by fuqianggao** [comment]
Thank you @mattklein123 and @htuch for reviewing this!

@htuch, can you remove the "requested changes" so this can be merged?

**Comment by fuqianggao** [comment]
@htuch @mattklein123 Made a minor change: make MetadataMatcher inside NodeMatcher to enable match on multiple metadata fields. PTAL again. Thanks!

File: api/envoy/type/matcher/v3/node.proto [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+// Specifies the way to match a Node.
+message NodeMatcher {
+  option (udpa.annotations.versioning).previous_message_type = "envoy.type.matcher.NodeMatcher";
+
+  // Specifies match criteria on the node id.
+  StringMatcher node_id = 1;
+
+  // Specifies match criteria on the node metadata.
+  repeated MetadataMatcher node_metadatas = 2;
```
Comment (htuch) [inline api/envoy/type/matcher/v3/node.proto]: Is this `AND` semantics? Might it be the case that someone one day wants `OR`?

File: api/envoy/service/status/v3/csds.proto [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+  ERROR = 4;
+}
+
+// Request for client status of clients identified by a list of NodeMatchers.
+message ClientStatusRequest {
+  option (udpa.annotations.versioning).previous_message_type =
+      "envoy.service.status.v2.ClientStatusRequest";
+
+  // Management server can use these match criteria to identify clients.
+  repeated type.matcher.v3.NodeMatcher node_matchers = 1;
```
Comment (htuch) [inline api/envoy/service/status/v3/csds.proto]: Do we need multiple of these if `NodeMatcher` is sufficiently expressive?

File: api/envoy/type/matcher/v3/node.proto (hunk 14 above)
Comment (fuqianggao) [reply api/envoy/type/matcher/v3/node.proto]: Yes, this is intended to be AND. If OR is needed, one can have repeated NodeMatcher, and only include either node_id or node_metadatas in them.

File: api/envoy/service/status/v3/csds.proto (hunk 15 above)
Comment (fuqianggao) [reply api/envoy/service/status/v3/csds.proto]: I don't see how we could achieve requesting client status for multiple unique node ids with only a single NodeMatcher. For example, if I want to request client status for "client-abc", and "client-123", it's hard to come up with a match that will only select this two.

Comment (mattklein123) [reply api/envoy/service/status/v3/csds.proto]: So this is an OR match, right? Can you clarify this in the docs?

File: api/envoy/type/matcher/v3/node.proto (hunk 14 above)
Comment (mattklein123) [reply api/envoy/type/matcher/v3/node.proto]: Can you clarify this in the docs per my other comment both here and there?

File: api/envoy/type/matcher/node.proto [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+package envoy.type.matcher;
+
+import "envoy/type/matcher/metadata.proto";
+import "envoy/type/matcher/string.proto";
+
+option java_package = "io.envoyproxy.envoy.type.matcher";
+option java_outer_classname = "NodeProto";
+option java_multiple_files = true;
+
+// [#protodoc-title:NodeMatcher]
```
Comment (mattklein123) [inline api/envoy/type/matcher/node.proto]: nit: "// [#protodoc-title: Node matcher]"

Comment (fuqianggao) [reply api/envoy/type/matcher/node.proto]: I see that other files still have the original format, e.g.: https://github.com/envoyproxy/envoy/blob/master/api/envoy/type/matcher/v3/metadata.proto#L15

File: api/envoy/type/matcher/v3/node.proto (hunk 14 above)
Comment (fuqianggao) [reply api/envoy/type/matcher/v3/node.proto]: Done. Thank you.

File: api/envoy/service/status/v3/csds.proto (hunk 15 above)
Comment (fuqianggao) [reply api/envoy/service/status/v3/csds.proto]: Done.

File: api/envoy/type/matcher/node.proto (hunk 16 above)
Comment (mattklein123) [reply api/envoy/type/matcher/node.proto]: Do you mind fixing the other one(s)? It's a doc nit but generally we try for human readable docs.

**Comment by fuqianggao** [comment]
> LGTM other than small doc nit, thanks.
> 
> /wait

Done. Thanks for your review!
